From 168969c4de0a636ff0b1b86c9f25010496681e67 Mon Sep 17 00:00:00 2001
From: Laurent Pinchart <laurent.pinchart@ideasonboard.com>
Date: Fri, 28 Nov 2014 10:35:57 +0200
Subject: [PATCH 577/827] v4l: xilinx: dma: Return buffers to vb2 when media pipeline start fails

This patch comes from:
  https://github.com/Xilinx/linux-xlnx.git

The start_streaming operation must return all queued buffers to vb2 in
case of failure. This is already implemented except in case of media
pipeline start failures. Fix it.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>
(cherry picked from commit 6678d9c6e8e95a87dcd701ed1fecc011c4b3d6bb)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/media/platform/xilinx/xilinx-dma.c | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/drivers/media/platform/xilinx/xilinx-dma.c b/drivers/media/platform/xilinx/xilinx-dma.c
index 0b276a8..d54e4c1 100644
--- a/drivers/media/platform/xilinx/xilinx-dma.c
+++ b/drivers/media/platform/xilinx/xilinx-dma.c
@@ -431,18 +431,18 @@ static int xvip_dma_start_streaming(struct vb2_queue *vq, unsigned int count)
 
 	ret = media_entity_pipeline_start(&dma->video.entity, &pipe->pipe);
 	if (ret < 0)
-		return ret;
+		goto error;
 
 	/* Verify that the configured format matches the output of the
 	 * connected subdev.
 	 */
 	ret = xvip_dma_verify_format(dma);
 	if (ret < 0)
-		goto error;
+		goto error_stop;
 
 	ret = xvip_pipeline_prepare(pipe, dma);
 	if (ret < 0)
-		goto error;
+		goto error_stop;
 
 	/* Start the DMA engine. This must be done before starting the blocks
 	 * in the pipeline to avoid DMA synchronization issues.
@@ -454,9 +454,10 @@ static int xvip_dma_start_streaming(struct vb2_queue *vq, unsigned int count)
 
 	return 0;
 
-error:
+error_stop:
 	media_entity_pipeline_stop(&dma->video.entity);
 
+error:
 	/* Give back all queued buffers to videobuf2. */
 	spin_lock_irq(&dma->queued_lock);
 	list_for_each_entry_safe(buf, nbuf, &dma->queued_bufs, queue) {
-- 
2.9.3

